### PR TITLE
[6.0] GetVar(inputIndex, varId) for reading context variable from another input and fix for Context.getVar

### LINKED ISSUE #978: Context.getVar failing compilation (and likely can't be deserialized
test for method-call version of `getVar` (you can put it in `BasicOpsSpecification`) is failing compilation
```
property("getVar") {
    test("getVar", env, ext,
      "{ allOf(Coll(CONTEXT.getVar[Boolean](trueVar).get, true, true)) }",
      AND(GetVarBoolean(booleanVar).get, TrueLeaf, TrueLeaf).toSigmaProp
    )
  }
```
(and all the method call with parametrized types as well)

Also, after fixing compilation serialization will be failed for `Context.getVar[T]` , as `MethodCallSerializer` is not serializing and so not able to restore `T`

### PR DESCRIPTION
**Pull request by kushti**
This PR contains: 

* fix for CONTEXT.getVar failing deserialization and compilation ( close #978 )
* CONTEXT.getVarFromInput method to read context variable from any input.
 
  Note, that the new method is intentionally not throwing an exception on expected vs real types. This can be used to have `getVar` alternative without an exception, as in:

```
     {
         val idx = CONTEXT.selfBoxIndex
         sigmaProp(CONTEXT.getVarFromInput[Int](idx.toShort, 1.toByte).get == 5)
     }
```

### CODE REVIEW

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 1]
```diff
@@ -1522,6 +1527,148 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
     )
   }
 
+  property("Context.getVar and getVarFromInput") {
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1530]: Better use separate `property` for each method.

File: sc/shared/src/test/scala/sigmastate/lang/SigmaTyperTest.scala [hunk 2]
```diff
@@ -683,6 +683,10 @@ class SigmaTyperTest extends AnyPropSpec
     typecheck(env, "CONTEXT.dataInputs") shouldBe SCollection(SBox)
   }
 
+  property("SContext.getVar") {
+    typecheck(env, "CONTEXT.getVar[Int](1.toByte).get") shouldBe SInt
+  }
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigmastate/lang/SigmaTyperTest.scala:688]: A property for getVarFromInput is missing.

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala [hunk 3]
```diff
@@ -144,6 +147,77 @@ class BasicOpsSpecification extends CompilerTestingCommons
...
+        null
+      )
+    }
+
+    if(VersionContext.current.isV6SoftForkActivated) {
+      varTest()
+    } else {
+      an[Exception] should be thrownBy(varTest())
+    }
+  }
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala:163]: Missing test case for different actual and expected types.

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala [hunk 4]
```diff
@@ -144,6 +147,77 @@ class BasicOpsSpecification extends CompilerTestingCommons
...
+        null
+      )
+    }
+
+    if (VersionContext.current.isV6SoftForkActivated) {
+      getVarTest()
+    } else {
+      an[Exception] should be thrownBy getVarTest()
+    }
+  }
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala:220]: Missing test case for different actual and expected types.

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala [hunk 5]
```diff
@@ -1734,4 +1804,26 @@ class BasicOpsSpecification extends CompilerTestingCommons
...
+
+    test("R2", env, ext,
+      script = "", /* means cannot be compiled
+                   the corresponding script is "{ SELF.R4[Unit].get == () }" */
+      EQ(ExtractRegisterAs[SUnit.type](Self, reg1)(SUnit).get, UnitConstant.instance).toSigmaProp,
+      additionalRegistersOpt = Some(Map(
+        reg1 -> UnitConstant.instance
+      ))
+    )
+  }
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala:1827]: What this test means?

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala (hunk 5 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala:1862]: how is this test related to the PR ? Also, why to ask others about tests introduced by you (starting from  [4f2579bbaeec )3def75d42629d7821073f39804b2](https://github.com/ergoplatform/sigmastate-interpreter/commit/4f2579bbaeec3def75d42629d7821073f39804b2)  )

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala (hunk 1 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1533]: splitted

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala (hunk 3 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala]: Reworked into more nuanced check (for sigma.validation.ValidationException now)

File: sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala (hunk 4 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigmastate/utxo/BasicOpsSpecification.scala:207]: Reworked into more nuanced check (for sigma.validation.ValidationException now)

File: core/shared/src/main/scala/sigma/SigmaDsl.scala [hunk 6]
```diff
@@ -589,6 +589,17 @@ trait Context {
...
   def getVar[T](id: Byte)(implicit cT: RType[T]): Option[T]
 
+  /**
+    * A variant of `getVar` to extract a context variable by id and type from any input
+    *
+    * @param inputId - input index
+    * @param id - context variable id
+    * @tparam T - expected type of the variable
+    * @return Some(value) if the variable is defined in the context AND has the given type.
+    *         None otherwise
```
[the remark this answers is not in the document]
Comment (kushti) [reply core/shared/src/main/scala/sigma/SigmaDsl.scala:599]: I guess it was your proposal to relax typechecking (cant find it , probably was just done over a chat). However, for existing getVar, changing semantics is not dangerous. However, we can have relaxation via this new method (see PR description). But I am open to alternatives.

**Comment by kushti** [comment]
@aslesarenko comments addressed , please make another pass

File: core/shared/src/main/scala/sigma/SigmaDsl.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/SigmaDsl.scala:599]: yeah, make sense.

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**